.github/workflows/nestbuildmatrix.yml

```diff
@@ -1,167 +1,368 @@
+            os: "ubuntu-20.04"
+          - xNEST_BUILD_TYPE: "OPENMP_ONLY"
+            cpp_compiler: "clang"
+            os: "ubuntu-20.04"
+          - xNEST_BUILD_TYPE: "FULL"
+            os: "ubuntu-20.04"
+            cpp_compiler: "clang"
+
+    steps:
+      - name: "Checkout repository content"
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+
+      - name: "Find changed files"
+        run: |
+          echo "CHANGED_FILES<<EOF" >>$GITHUB_ENV
+          echo "$(git diff --name-only ${{ github.event.before }}..${{ github.event.after }})" >>$GITHUB_ENV
+          echo 'EOF' >>$GITHUB_ENV
+
+      - name: "DEBUG: Check path"
+        run: |
+          echo $PATH
+          echo $PWD
+        shell: bash
+
+      - name: "Set up Python 3.x"
+        uses: actions/setup-python@v2
+        with:
+          python-version: 3.9
+
+      - name: "DEBUG: Check Python path"
+        run: echo $PYTHONPATH
+
+      - name: "Restore apt cache"
+        uses: actions/cache@v2
+        env:
+          cache-name: "apt-cache"
+        with:
+          path: |
+            !/var/cache/apt/archives/lock
+            !/var/cache/apt/archives/partial
+            /var/cache/apt
+          key: ${{ runner.os }}-system-${{ env.cache-name }}-${{ hashFiles('**/environment.yml') }}
+          restore-keys: |
+            ${{ runner.os }}-system-${{ env.cache-name }}-
+            ${{ runner.os }}-system-
+
+      - name: "Install Linux system dependencies"
+        run: |
+          sudo apt-get update
+          #https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md
+          sudo apt-get install libltdl-dev libreadline6-dev libncurses5-dev libgsl0-dev python3-all-dev jq pycodestyle libpcre3 libpcre3-dev python2-dev libboost-all-dev
+          sudo apt-get install openmpi-bin libopenmpi-dev libgsl0-dev tcl8.6 tcl8.6-dev tk8.6-dev
+          sudo apt-get install libboost-filesystem-dev libboost-regex-dev libboost-wave-dev libboost-python-dev libboost-program-options-dev libboost-test-dev
+          sudo apt-get install vera++
+          sudo ldconfig
+          g++ --version
+
+      - name: "Restore pip cache"
+        env:
+          cache-name: "pip-cache"
+        uses: actions/cache@v2
+        with:
+          path: |
+            /opt/hostedtoolcache/Python/**/site-packages/*
+          key: ${{ runner.os }}-python-${{ env.cache-name }}-${{ hashFiles('**/requirements.txt') }}
+          restore-keys: |
+            ${{ runner.os }}-python-${{ env.cache-name }}-
+            ${{ runner.os }}-python-
+
+      - name: "Install Python dependencies"
+        run: |
+          python -m pip install --upgrade pip setuptools
+          python -c "import setuptools; print('package location:', setuptools.__file__)"
+          python -m pip install --force-reinstall --upgrade scipy 'junitparser>=2' numpy pytest pytest-timeout pytest-xdist mpi4py cython matplotlib terminaltables pandoc
```

Comment:
Because the packages that are already installed in the default image are broken shims. For example `cython` is installed, but can not `cythonize` code. – Took some days to figure out.

Reply:
Does caching add any benefit if we force reinstall every time?

Reply:
After some discussions we had, I'd opt for leaving it in for now. I don't see a clean way to distinguish if the installed packages are the fixed ones from the cache, or the broken default ones. In any case, having the cache rebuilt, will already have all the dependencies already, so the `--force-reinstall` does not traverse the whole dependency tree. The `--force-reinstall` needs to stay in because otherwise a cache-miss would cause a broken build. There is a separate cache for pip downloads, that I added now hoping to limit externally required resources even further. The exact cache location can be found with `pip cache dir`, but I'm not sure how to have that expanded in the GHA.

Reply:
Thanks for the clarification!